scripts/react-compiler-compliance-check.ts

```diff
@@ -30,55 +44,78 @@ type CompilerFailure = {
     file: string;
     line?: number;
     column?: number;
-    reason: string;
+    reason?: string;
 };
 
-function check(filesToCheck?: string[], shouldGenerateReport = false) {
+type DiffFilteringCommits = {
+    from: string;
+    to: string;
+};
+
+type PrintResultsOptions = {
+    shouldPrintSuccesses: boolean;
+};
+
+type CommonCheckOptions = PrintResultsOptions & {
+    remote?: string;
+    reportFileName?: string;
+    shouldGenerateReport?: boolean;
+    shouldFilterByDiff?: boolean;
+};
+
+type CheckOptions = CommonCheckOptions & {
+    filesToCheck?: string[];
+};
+
+async function check({
+    filesToCheck,
+    shouldGenerateReport = false,
+    reportFileName = DEFAULT_REPORT_FILENAME,
+    shouldFilterByDiff = false,
+    remote,
+    shouldPrintSuccesses = false,
+}: CheckOptions): Promise<boolean> {
     if (filesToCheck) {
-        info(`Running React Compiler check for ${filesToCheck.length} files or glob patterns...`);
+        logInfo(`Running React Compiler check for ${filesToCheck.length} files or glob patterns...`);
     } else {
-        info('Running React Compiler check for all files...');
+        logInfo('Running React Compiler check for all files...');
     }
 
     const src = createFilesGlob(filesToCheck);
-    const results = runCompilerHealthcheck(src);
+    let results = runCompilerHealthcheck(src);
 
-    const isPassed = results.failures.size === 0;
-    if (isPassed) {
-        logSuccess('All changed files pass React Compiler compliance check!');
-        return true;
+    if (shouldFilterByDiff) {
+        const mainBaseCommitHash = await Git.getMainBranchCommitHash(remote);
+        const headCommitHash = 'HEAD';
+        const diffFilteringCommits: DiffFilteringCommits = {from: mainBaseCommitHash, to: headCommitHash};
+        results = await filterResultsByDiff(results, diffFilteringCommits, {shouldPrintSuccesses});
     }
 
-    printFailureSummary(results);
+    printResults(results, {shouldPrintSuccesses});
 
     if (shouldGenerateReport) {
-        generateReport(results, DEFAULT_REPORT_FILENAME);
+        generateReport(results, reportFileName);
     }
 
-    return false;
+    const isPassed = results.failures.size === 0;
+    return isPassed;
 }
 
-async function checkChangedFiles(remote: string): Promise<boolean> {
-    info('Checking changed files for React Compiler compliance...');
-
-    try {
-        const changedFiles = await Git.getChangedFiles(remote);
-        const filesToCheck = [...new Set(changedFiles)];
+type CheckChangedFilesOptions = CommonCheckOptions;
```

Comment:
Why have a duplicate type alias?

Reply:
I though i'd keep it to make the type for the options of `checkChanged` more explicit. Will remove!